Fetch w/o CB if mempool empty, don't use HB mode if blocks only.

gmaxwell opened the pull request:
Fetching with CB when there are less than 30 hits for a full block
 wastes bandwidth; and getting three copies of a CB message which
 is of no use to us when we are in blocks only mode is not great.

This is a fairly small optimization, since even the worst case of
 getting three useless CB messages is only a 5% bandwidth increase.

[I will update the test if this gets Concept ACKs]

laanwj commented:
Concept ACK

jonasschnelli commented:
Concept ACK, needs rebase.

sipa commented:
Concept ACK, needs rebase.

gmaxwell commented:
Rebased, I still need to fix the tests.

rebroad commented on src/main.cpp:
```diff
@@ -5341,7 +5341,7 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                         (!IsWitnessEnabled(chainActive.Tip(), chainparams.GetConsensus()) || State(pfrom->GetId())->fHaveWitness)) {
                         inv.type |= nFetchFlags;
-                        if (nodestate->fSupportsDesiredCmpctVersion)
+                        if (nodestate->fSupportsDesiredCmpctVersion && mempool.size() > 1)
```
is 1 sufficient?

rebroad commented on src/main.cpp:
```diff
@@ -6008,10 +6008,13 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-                        MaybeSetPeerAsAnnouncingHeaderAndIDs(nodestate, pfrom, connman);
+                        // with this block! Let's get them to announce using compact blocks in the future, unless
+                        // either of us are blocks-only and would prefer to save bandwidth.
+                        if (pfrom->fRelayTxes && !GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY)) {
```
Why use two different tests? I'd have though the mempool size test for both, or the blockonly test for both would bring more consistency.

sdaftuar replied:
I'm not sure why we wouldn't ask this peer for compact blocks just because they are not relaying transactions.  Our mempool might be perfectly good for block reconstruction, even if the peer who is relaying blocks to us fastest is blocksonly, no? 

I know that in the future we're likely to want to optimize the prefilled transactions based eg on what was missing from our own mempools, but I think at the time we do that we might as well start scoring our peers based on, say, how often a roundtrip is required or something.  And if a blocksonly peer happens to be really fast to deliver us blocks then great...

gmaxwell replied:
The thinking there was that if the peer is trying to economize bandwidth, we shouldn't ask them to send the unsolicited compact blocks (HB mode). We'll still request compact blocks from them when advertised.  I don't fee super strongly, if they're trying to economize on bandwidth they should probably delay their relays artificially.  We don't have a NAK for HB mode requests though.

gmaxwell committed:
Fetch w/o CB if mempool empty, don't use HB mode if blocks only.

Fetching with CB when there are less than 30 hits for a full block
 wastes bandwidth; and getting three copies of a CB message which
 is of no use to us when we are in blocks only mode is not great.

This is a fairly small optimization, since even the worst case of
 getting three useless CB messages is only a 5% bandwidth increase.

rebroad commented on src/net_processing.cpp:
```diff
@@ -2095,10 +2095,13 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
-                        MaybeSetPeerAsAnnouncingHeaderAndIDs(nodestate, pfrom, connman);
+                        // with this block! Let's get them to announce using compact blocks in the future, unless
+                        // either of us are blocks-only and would prefer to save bandwidth.
+                        if (pfrom->fRelayTxes && !GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY)) {
```
Why not put this check within MaybeSetPeerAsAnnouncingHeaderAndIDs() ?
